windows-installer/auto-update.ps1

```diff
@@ -0,0 +1,134 @@
+        $certbotSigningPubKey = (Get-ItemProperty -Path HKLM:\Software\Certbot).CertbotSigningPubKey
+    } else {
+        $certbotSigningPubKey = '
+-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA6MR8W/galdxnpGqBsYbq
+OzQb2eyW15YFjDDEMI0ZOzt8f504obNs920lDnpPD2/KqgsfjOgw2K7xWDJIj/18
+xUvWPk3LDkrnokNiRkA3KOx3W6fHycKL+zID7zy+xZYBuh2fLyQtWV1VGQ45iNRp
+9+Zo7rH86cdfgkdnWTlNSHyTLW9NbXvyv/E12bppPcEvgCTAQXgnDVJ0/sqmeiij
+n9tTFh03aM+R2V/21h8aTraAS24qiPCz6gkmYGC8yr6mglcnNoYbsLNYZ69zF1XH
+cXPduCPdPdfLlzVlKK1/U7hkA28eG3BIAMh6uJYBRJTpiGgaGdPd7YekUB8S6cy+
+CQIDAQAB
+-----END PUBLIC KEY-----
+'
+    }
+
+    if ((Test-Path HKLM:\Software\Certbot) -And ((Get-ItemProperty -Path HKLM:\Software\Certbot).PSObject.Properties.Name -Contains "CertbotUpgradeApiURL")) {
+        $certbotUpgradeApiURL = (Get-ItemProperty -Path HKLM:\Software\Certbot).CertbotUpgradeApiURL
+    } else {
+        $certbotUpgradeApiURL = 'https://api.github.com/repos/certbot/certbot/releases/latest'
+    }
+
+    # Get current local certbot version
+    try {
+        $currentVersion = certbot --version
+        $currentVersion = $currentVersion -replace '^certbot (\d+\.\d+\.\d+).*$', '$1'
+        $currentVersion = [System.Version]"$currentVersion"
+    } catch {
+        Write-Message @"
+An error occured while fetching the current local certbot version:
+$_
+Assuming Certbot is not up-to-date.
+"@ "Warning"
+        $currentVersion = "0.0.0"
+    }
+
+    # Get latest remote certbot version
+    try {
+        [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
+        $result = Invoke-RestMethod -Uri $certbotUpgradeApiURL
+        $latestVersion = $result.tag_name -replace '^v(\d+\.\d+\.\d+).*$', '$1'
+        $latestVersion = [System.Version]"$latestVersion"
+    } catch {
+        Write-Error @"
+Could not get the latest remote certbot version. Error was:
+$_
+Aborting auto-upgrade process.
+"@
+    }
+
+    if ($currentVersion -ge $latestVersion) {
+        Write-Message "No upgrade is needed, Certbot is already at the latest version ($currentVersion)."
+    } else {
+        # Search for the Windows installer asset
+        $installerUrl = $null
+        foreach ($asset in $result.assets) {
+            if ($asset.name -match '^certbot-.*installer-win32\.exe$') {
+                $installerUrl = $asset.browser_download_url
+            }
+        }
+
+        if ($null -eq $installerUrl) {
+            Write-Error "Could not find the URL for the latest Certbot for Windows installer."
+        }
+
+        Write-Message "Starting Certbot auto-upgrade from $currentVersion to $latestVersion ..."
+
+        $installerPath = "$env:TMP/certbot-installer-win32.exe"
+        try {
+            # Download the installer
+            Write-Message "Downloading the installer ..."
+            $webClient = New-Object System.Net.WebClient
```

Comment:
Indeed, we can reuse Invoke-RestMethod here. I am not convinced by the wrapper for two calls, since it requires boilerplate code to handle the very different situations here (REST API call and binary download). I added a timeout however.